feat: reusable containers

matthiasschaub opened the pull request:
adresses #109

Todo:
- [x] Write documentation about how to reuse containers.
- [x] Warn the user if `with_reuse` in use but ryuk is disabled.
- [x] Address https://github.com/testcontainers/testcontainers-python/pull/636/changes#r1698932279

Open questions:
- [ ] Should `reuse_enable` also be configurable via environment variable?
- [ ] Currently to make reuse work ryuk needs to be disabled. Should the user do this manually? -> if so warn the user if `with_reuse` in use but ryuk is disabled.
- [ ] Current tests are sensitive to users `~/.testcontainers.properties`. This file should not be present during test run. 
- [ ] Should each container class have the possibility to state if this container can be reused? (`DockerContainer.reusable: bool = True`)

matthiasschaub committed:
feat: reusable containers

adresses #109

Co-authored-by: Levi Szamek <levi.szamek@heigit.org>

matthiasschaub commented on index.rst:
```diff
@@ -85,6 +85,29 @@ When trying to launch a testcontainer from within a Docker container, e.g., in c
+    ...     second_id == container._container.id
+    >>> print(first_id == second_id)
+    True
+
```
Should the user be warned that by using this feature, containers need to be removed manually? (That this feature should not be used in a CI)

Also, do we need to make clear how this feature works (explaining the hash in use). -> If a container's run configuration changes, the hash changes and a new container will be used.

alexanderankin replied:
seems like you have added these comments to the doc, i think that is fine. the hash would be great to add as users would benefit from knowing exactly what is hashed.

* `self.image,`
* `self._command,`
* `self.env,`
* `self.ports,`
* `self._name,`
* `self.volumes,`
* `str(tuple(sorted(self._kwargs.items()))),` - this may fail and why i want to have this be tucked away inside an obviously readable `if` block

matthiasschaub replied:
Maybe use something like https://github.com/knowsuchagency/picocache/blob/main/picocache/utils.py#L9 for making the hash key

matthiasschaub replied:
Before running `str(tuple(sorted(self._kwargs.items())))`, `self._kwargs` is now beeing checked for `None`

matthiasschaub commented on core/testcontainers/core/config.py:
```diff
@@ -73,8 +76,19 @@ def docker_auth_config(self, value: str):
+    def tc_properties_tc_host(self) -> Union[str, None]:
+        return self.tc_properties.get("tc.host")
+
+    @property
+    def tc_properties_testcontainers_reuse_enable(self) -> bool:
+        enabled = self.tc_properties.get("testcontainers.reuse.enable")
+        return enabled == "true"
```
`testcontainers.reuse.enable` (like the Java implementation) or `tc.reuse.enable` (like `tc.host`)?

`enable` (like the Java implementation) or `enabled` (like RYUK_DISABLED)?

eddumelendez replied:
Let's keep the same key as Java implementation `testcontainers.reuse.enable`.

kiview replied:
Yes, this is important. This key is used across most language implementations and the way tc-java does is considered the reference implemetation.

matthiasschaub replied:
Okay, then let's keep it as implemented.

matthiasschaub committed:
docs: add documentation about reusable containers

matthiasschaub committed:
test: additional testcase for reusable containers

matthiasschaub committed:
test: add newlines for better readability

matthiasschaub committed:
warn user if ryuk is disabled but with_reuse used

matthiasschaub committed:
docs: fix code highlighting

matthiasschaub committed:
fix: use Union instead of | for type hint

alexanderankin commented:
will review and merge if you have no other plans for changes

matthiasschaub commented:
@alexanderankin we would welcome a review from you, thanks. They are no further plans right now except addressing what ever comes up in the code review.

matthiasschaub committed:
refactor: remove TODO comment

matthiasschaub committed:
docs: update section on reusable containers

alexanderankin commented on core/testcontainers/core/container.py:
```diff
@@ -86,8 +93,49 @@ def start(self) -> Self:
         self._configure()
+
+        # container hash consisting of run arguments
+        args = (
+            self.image,
+            self._command,
+            self.env,
+            self.ports,
+            self._name,
+            self.volumes,
+            str(tuple(sorted(self._kwargs.items()))),
+        )
+        hash_ = hashlib.sha256(bytes(str(args), encoding="utf-8")).hexdigest()
+
+        if self._reuse and (not c.tc_properties_testcontainers_reuse_enable or not c.ryuk_disabled):
+            logging.warning(
+                "Reuse was requested (`with_reuse`) but the environment does not "
+                + "support the reuse of containers. To enable container reuse, add "
+                + "the 'testcontainers.reuse.enable=true' to "
+                + "'~/.testcontainers.properties' and disable ryuk by setting the "
+                + "environment variable 'TESTCONTAINERS_RYUK_DISABLED=true'"
+            )
+
+        if self._reuse and c.tc_properties_testcontainers_reuse_enable:
+            docker_client = self.get_docker_client()
+            container = docker_client.find_container_by_hash(hash_)
+            if container:
+                if container.status != "running":
+                    container.start()
+                    logger.info("Existing container started: %s", container.id)
+                logger.info("Container is already running: %s", container.id)
+                self._container = container
+            else:
+                self._start(hash_)
+        else:
+            self._start(hash_)
```
can we refactor this so that is is obvious where the if clause is that triggers this?

want to make sure
1. we are doing the hash inside the clause
2. want to make it more readable - not in general but specifically for ensuring correcteness of logic that disables or enables reuse

matthiasschaub replied:
Thanks for the feedback! I will revisit this part next week and try to improve upon it.

matthiasschaub replied:
@alexanderankin I moved the generation of the hash inside the if-clause and removed passing the `hash_` to start if *reuse* is not in use. I think that makes it better readable in general.

alexanderankin commented on core/testcontainers/core/container.py:
```diff
@@ -96,16 +144,17 @@ def start(self) -> Self:
             ports=self.ports,
             name=self._name,
             volumes=self.volumes,
+            labels={"hash": hash_},
             **self._kwargs,
         )
         logger.info("Container started: %s", self._container.short_id)
-        if self._network:
-            self._network.connect(self._container.id, self._network_aliases)
-        return self
 
     def stop(self, force=True, delete_volume=True) -> None:
         if self._container:
-            self._container.remove(force=force, v=delete_volume)
+            if self._reuse and c.tc_properties_testcontainers_reuse_enable:
+                self._container.stop()
+            else:
+                self._container.remove(force=force, v=delete_volume)
```
hm, isnt the point to not even stop it so it is warm for next run? i guess if people are using the explicit api then whatever. I do see a bit of a mirror with start so i guess it will just have to be consistent and maybe clear in docs.

kiview replied:
In other languages, having a reusable container *does not* change the contract of the `stop()` method. This is obviously something that needs to be considered to make this a full fledged use case, but as of now, I would suggest we start with an experimental reusable implementation, that mirrors the Java implementation.

matthiasschaub replied:
Okay, I see. I updated the code and documentation (how to use reusable containers) to not change the contract of the `stop()` method.

alexanderankin commented on index.rst:
```diff
@@ -118,6 +117,32 @@ Fetching passwords from cloud providers:
     GCP_PASSWORD = $(gcloud auth print-access-token)
     AZURE_PASSWORD = $(az acr login --name <registry-name> --expose-token --output tsv)
 
+Reusable Containers (Experimental)
+----------------------------------
+
+Containers can be reused across consecutive test runs. To reuse a container, the container configuration must be the same.
+
+Containers that are set up for reuse will not be automatically removed. Thus, those containers need to be removed manually.
```
"...removed manually."

maybe add:
"In re-usable mode, the 'stop' api on a container will now 'stop' a container, rather than 'remove' it"

matthiasschaub replied:
After [this](https://github.com/testcontainers/testcontainers-python/pull/636#discussion_r1701699697) discussion, the stop method has not been changed.

alexanderankin commented on core/testcontainers/core/container.py:
```diff
@@ -86,8 +93,49 @@ def start(self) -> Self:
+        else:
+            self._start(hash_)
+
+        if self._network:
+            self._network.connect(self._container.id, self._network_aliases)
+        return self
```
apparently this part is also fairly jank and we should remove/rework so as a note to myself i can only do that after this pr merges

kiview commented:
Thanks for looking into this @matthiasschaub 👋 
We have no clean spec for this features in Java / across languages, so for now I would suggest, we mostly mirror the Java implementation, including its limitations. We can sync on a cross language spec in the future that provides a better DX, but for now I would strongly favor an implementation that mirrors Java.

I already left some comments within the PR.

> Currently to make reuse work ryuk needs to be disabled. Should the user do this manually?

Given the above, if a container has `reusable` set, it *must not* be registered with the Ryuk cleanup label (see [tc-java](https://github.com/testcontainers/testcontainers-java/blob/c491f6aa543a4657e53578e8a0581ab375f8c105/core/src/main/java/org/testcontainers/containers/GenericContainer.java#L432-L435)).

kiview commented on core/testcontainers/core/container.py:
```diff
@@ -86,8 +93,49 @@ def start(self) -> Self:
+            self.volumes,
+            str(tuple(sorted(self._kwargs.items()))),
+        )
+        hash_ = hashlib.sha256(bytes(str(args), encoding="utf-8")).hexdigest()
```
Ideally we use the *full* container create request as the hash input. In tc-java, this is the `CreateContainerCmd` from docker-java, I guess we have some equivalent request object from the Docker Python SDK somewhere available?

matthiasschaub replied:
Yeah, that would be the ideal solution. Unfortunately, I could not find an equivalent function to `CreateContainerCmd` in the Docker SDK for Python: Not by going through the documentation and not by browsing the code base.

matthiasschaub replied:
Instead of `sh256` its probably cheaper and faster to use `zlib.crc32`.

matthiasschaub committed:
feat(reuse): do not change contract of stop method

matthiasschaub committed:
feat(reuse): do not create Ryuk cleanup instance

do not create Ryuk cleanup instance if reuse enabled and container has
been start with `with_reuse`

matthiasschaub commented:
> Thanks for looking into this @matthiasschaub 👋 We have no clean spec for this features in Java / across languages, so for now I would suggest, we mostly mirror the Java implementation, including its limitations. We can sync on a cross language spec in the future that provides a better DX, but for now I would strongly favor an implementation that mirrors Java.
> 
> I already left some comments within the PR.
> 
> > Currently to make reuse work ryuk needs to be disabled. Should the user do this manually?
> 
> Given the above, if a container has `reusable` set, it _must not_ be registered with the Ryuk cleanup label (see [tc-java](https://github.com/testcontainers/testcontainers-java/blob/c491f6aa543a4657e53578e8a0581ab375f8c105/core/src/main/java/org/testcontainers/containers/GenericContainer.java#L432-L435)).

Thanks for the review! I agree. It is sensible to follow the Java implementation.

In commit 1ea9ed16a0dd73e11d762808aed5655d44e270be I do not create a Reaper instance during container start-up if reuse is enabled and container has been started with `with_reuse`. This works as expected as long as no other container is started without `with_reuse`. In that case a Reaper instances is created which will remove the reuse container as well.
Is there a way to explicitly exclude a container from the Reapers routine?

matthiasschaub committed:
refactor: move hash generation into if clause

matthiasschaub committed:
Merge remote-tracking branch 'origin/main' into reusable_containers

matthiasschaub committed:
Merge branch 'main' into reusable_containers

matthiasschaub commented:
Please let me know if I can do anything to advance this PR towards merging :)

matthiasschaub committed:
Merge branch 'main' into reusable_containers

matthiasschaub committed:
Merge branch 'main' into reusable_containers

gpkc commented on core/testcontainers/core/config.py:
```diff
@@ -125,9 +128,19 @@ def docker_auth_config(self, value: str) -> None:
+        enabled = self.tc_properties.get("testcontainers.reuse.enable")
+        return enabled == "true"
+
     def timeout(self) -> int:
```
I just noticed that `timeout` here had the `@property` decorator before. Was it removed by accident?

matthiasschaub replied:
Yes, this was an accident.

matthiasschaub committed:
fix: revert accidental removal of @property

chbndrhnns commented:
Is this PR still intended to be merged?

matthiasschaub committed:
Merge remote-tracking branch 'testcontainers/main' into reusable_containers

matthiasschaub committed:
Merge remote-tracking branch 'testcontainers/main' into reusable_containers

matthiasschaub committed:
Merge branch 'main' into reusable_containers

matthiasschaub committed:
tests: access tc properties through function call

and assert not none to make mypy happy.

matthiasschaub committed:
refactor: use crc32 instead of sha1 to compute hash

To compute a hash to identify container across runs use crc32 instead of
sha1 which is faster while still being fit-for-purpose.

matthiasschaub committed:
refactor: remove remnant code from past merge

matthiasschaub committed:
docs: more detailed explanation on how to use reusable containers

matthiasschaub committed:
fix: check if self._kwargs is None before using it

matthiasschaub commented:
When reviewing, please also check open questions stated in the PR description. Thanks :)

dantetemplar commented:
Upvote

jacky-micro1 commented:
upvote

matthiasschaub committed:
Merge remote-tracking branch 'origin/main' into reusable_containers

matthiasschaub committed:
fix: restore deleted method after resolving merge conflict

matthiasschaub commented:
I incorporated latest changes from `main` into this branch. I would be curious if the CI runs through successfully.
